docs: guide for porting an html site to gatsby (#18609)
gatsbyjs/gatsby: docs/docs/porting-part-of-a-html-site-to-gatsby.md

### Gatsby Project

Now that you are setup we'll be using the Gatsby and NPM CLI tools on your command line to get this site section ported!
Make a new project using the Gatsby hello world starter with the following command:

```shell
gatsby new gatsby-site-section https://github.com/gatsbyjs/gatsby-starter-hello-world
```

You now have a folder called `gatsby-site-section` containing a basic Gatsby website. Open the new folder in your code editor and `cd` (change directory) into the folder in your command line to continue:

```shell
cd gatsby-site-section
```

The `/src` folder contains all code for the Gatsby site. In the Gatsby build stage every file in the `/src/pages` folder will result in a HTML file. Currently, the only page file is the root `/src/pages/index.js`:

```jsx:title=/gatsby-site-section/src/pages/index.js
import React from "react"

export default () => <div>Hello world!</div>
```

> Run the development server in the command line to see the website:

Author: Yeh I agree, I have a strong urge to keep 'em on the page but this is a good spot to remind them about the rest of the docs that they can move onto after too.

```jsx:title=/gatsby-site-section/src/pages/ella-arborist.js
import React from "react"
import Layout from "../components/Layout"
import { Link } from "gatsby"

export default () => (
  <Layout staffName="Ella">
    <h2>Ella is an excellent Arborist. We guarantee it.</h2>
    <div className="bio-card">
      <img
        alt="Comically crude stick person sketch"
        src="https://gatsby-html-partial-assets.s3.eu-west-2.amazonaws.com/person.png"
      />
      <p>Ella</p>
    </div>
  </Layout>
)
```

The other 2 pages for Marin and Sam can now be made with a similar structure. Maybe you are even thinking about another component for the Bio Card!

## Build

With the implementation of the site section into Gatsby complete, it's time to build and integrate into the rest of the website. There are a couple of config items to take care of to ensure a ported Gatsby site section is seamlessly integrated.

### Path Prefix

> By default, Gatsby assumes that the build will be hosted at the root path of a domain, e.g `https://gatsbyjs.org/`. After the initial static HTML file is loaded, Gatsby reaches out for other generated JavaScript files in the same location in order to progressively enable enhancements for the user. In the example site, the path at which the Gatsby site will be hosted is `/who`. The gatsby `pathPrefix` option will ensure that the generated JavaSript files are available to load and the generated links from `index.js` to internal pages work correctly:

Reviewer: ```suggestion
By default, Gatsby assumes that the build will be hosted at the root path of a domain, e.g `https://gatsbyjs.org/`. After the initial static HTML file is loaded, Gatsby reaches out for other generated JavaScript files in the same location in order to [progressively enable enhancements](/docs/glossary/#progressive-enhancement) for the user. In the example site, the path at which the Gatsby site will be hosted is `/who`. The gatsby `pathPrefix` option will ensure that the generated JavaSript files are available to load and the generated links from `index.js` to internal pages work correctly:
```

---
title: Porting Part of a HTML Site to Gatsby
---

## Why partially port a HTML website?

When introducing Gatsby into an existing static HTML site, it might not be feasible to rewrite the entire site at once. In these cases, porting pieces of the site to Gatsby one at a time, while the rest of the site still uses HTML, might be preferable. This page will guide you through this process.

> **Note**: The aim of this guide is to take a shallow, focused path to porting part of a static HTML website to Gatsby. The full [Gatsby tutorial](/tutorial/) is more broad, with deeper dives on the core concepts and technologies.

## Why Gatsby?

> ### Web building blocks

Author: Very much agree, I've tried to help the reader focus in the rest of the page but this section would certainly make my mind wander. Pushing up a rewrite shortly, keen to hear your thoughts on it.

There are two common issues faced when porting part of a static site:

- Correctly prefixing link paths within the site.
- Hosting static assets (CSS, images etc.) on a separate domain from the HTML files.
  Gatsby has built in options to set consistent and functional paths for both pages and assets.

### Dynamic content

Gatsby gathers content from any number of sources, into a single input point for each page. Once ported, content extracted from the site is easy to maintain in isolation from code. You can choose from many content management platforms such as Prismic and Contentful. Gatsby will bring the content together for consumption in your pages.

## Getting Started

### Which part to port?

Here is the structure of an example static HTML/CSS website that this guide will walk through:

```
 asset-domain
 ├── favicon.ico
 ├── person.png
 ├── normalize.css
 └── style.css
website-domain
  ├── index.html
  ├── 404.html
  ├── about.html
  ├── contact.html
  ├── services
  │   ├── index.html
  │   ├── growing.html
  │   ├── cleaning.html
  │   ├── shrinking.html
  └── who
      ├── index.html
      ├── ellla-arborist.html
      ├── marin-leafer.html
      └── sam-surgeon.html
```

> The `/who` section of the site is a great candidate for porting as it is all within a single folder. Through this guide, you will develop the ported Gatsby section in isolation before integrating into the site.

Author: Agreed, pushing up this addition shortly.

### Assumptions

The example site uses global CSS files (`style.css` and `normalize.css`), more sophisticated styling structures can be accommodated but will not be covered here. No client side JavaScript (e.g jQuery etc.) is on the example site, Gatsby may conflict with client side JavaScript if it is not removed when porting.

### Development environment

> Gatsby generates websites and compiles applications, but also has a tool for local development. With some setup your computer can run generation and the local development environment. Follow the [Gatsby development environment tutorial](/tutorial/part-zero/) to get setup and then return here to continue.

Author: Agreed! It implied that generation and compilation are different but wasn't clear how/why. Your version clears up that confusion without disclosing too much complexity.

### Gatsby Project

> Now that you are setup we'll be using the Gatsby and NPM CLI tools on your command line to get this site section ported!

Author: Ack! Was sure I had internalized that rule, should have done a quick cmd-f.

Make a new project using the Gatsby hello world starter with the following command:

```shell
gatsby new gatsby-site-section https://github.com/gatsbyjs/gatsby-starter-hello-world
```

You now have a folder called `gatsby-site-section` containing a basic Gatsby website. Open the new folder in your code editor and `cd` (change directory) into the folder in your command line to continue:

```shell
cd gatsby-site-section
```

> The `/src` folder contains all code for the Gatsby site. In the Gatsby build stage [every file in the `/src/pages` folder will result in a HTML file](/docs/recipes/#creating-pages-automatically). Currently, the only page file is the root `/src/pages/index.js`:

Author: Ah yes, seeing an interesting skill I need to work on here: [progressively disclose complexity](https://www.gatsbyjs.org/docs/gatsby-core-philosophy/#progressively-disclose-complexity) while not being misleading.

```jsx:title=/gatsby-site-section/src/pages/index.js
import React from "react"

export default () => <div>Hello world!</div>
```

[Run the development server](/docs/quick-start/#start-development-server) with `gatsby develop` in the command line to see the website in your browser. Hello Gatsby! 👋

## Porting index.html

Here is `/who/index.html` from the example site structure above:

```html:title=/website-domain/who/index.html
<html lang="en">
  <head>
    <title>Taylor's Tidy Trees - Who We Are</title>
    <link
      href="https://gatsby-html-partial-assets.s3.eu-west-2.amazonaws.com/favicon.ico"
      rel="shortcut icon"
      type="image/x-icon"
    />
    <link
      rel="stylesheet"
      href="https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.1/normalize.min.css"
      integrity="sha256-l85OmPOjvil/SOvVt3HnSSjzF1TUMyT9eV0c2BzEGzU="
      crossorigin="anonymous"
    />
    <link
      rel="stylesheet"
      type="text/css"
      href="https://gatsby-html-partial-assets.s3.eu-west-2.amazonaws.com/style.css"
    />
  </head>
  <body>
    <header>

>       <h3 class="brand-color">Taylor's Tidy Trees</h3>

Author: Aha! Yes. This is really a logo and Deque uni says [best practice is to start main content with a h1](https://dequeuniversity.com/assets/html/jquery-summit/html5/slides/headings.html#best_practice_start_content_with_h1) So I'll styling this logo with a class and get a h1 on every page. Looking back the pages do all have a heading shaped gap.

        <li><a href="/who/sam-surgeon.html">Sam (Tree Surgeon)</a></li>
        <li><a href="/who/marin-leafer.html">Marin (Leafer)</a></li>
      </ul>
    </main>
  </body>
</html>
```

In the following sections, you'll convert this block of HTML into its equivalent code in Gatsby.

### Head elements

You might have noticed that the `/src/pages/index.js` file doesn't have a `<html>`, `<head>` or `<body>`. Gatsby makes a basic html structure for each page and places the `/src/pages/index.js` output into its body. More `<head>` elements and `<html>` attributes are added to the output page with a module called [React Helmet](https://github.com/nfl/react-helmet). React Helmet is added to a Gatsby project in the command line with NPM and then to the Gatsby config file:

```shell
npm install --save react-helmet gatsby-plugin-react-helmet
```

Gatsby projects have a config file at `/gatsby-config.js` where options can be specified and plugins added. Add a plugin line with `gatsby-plugin-react-helmet` to your config file:

```js:title=/gatsby-site-section/gatsby-config.js
/**
 * Configure your Gatsby site with this file.
 *
 * See: https://www.gatsbyjs.org/docs/gatsby-config/
 */

> module.exports = {

Author: Great, yeh wanted to keep it clear of GraphQL, agree with giving it a direct mention at the end.

  plugins: ["gatsby-plugin-react-helmet"], // highlight-line
}
```

Now you can import the `<Helmet>` component to the `index.js` file and place `<header>` & `<main>` elements for the existing HTML. Gatsby pages must have a single root parent so add a [React Fragment component](https://reactjs.org/docs/fragments.html) around them:

```jsx:title=/src/pages/index.js
import React from "react"
import Helmet from "react-helmet"

export default () => (
  <>
    <Helmet></Helmet>
    <header></header>
    <main>
      <div>Hello world!</div>
    </main>
  </>
)
```

> 

Author: Grand, I went back and forth on mentioning JSX directly, agree with bringing it in. Certainly in danger of misleading the reader about what javascript is if it's not mentioned.

```jsx:title=/gatsby-site-section/src/pages/ella-arborist.js
import React from "react"
import Layout from "../components/Layout"
import { Link } from "gatsby"

export default () => (
  {/* highlight-start */}
  <Layout staffName="Ella">
  {/* highlight-end */}
    <h2>Ella is an excellent Arborist. We guarantee it.</h2>
    <div className="bio-card">
      <img
        alt="Comically crude stick person sketch"
        src="https://gatsby-html-partial-assets.s3.eu-west-2.amazonaws.com/person.png"
      />
      <p>Ella</p>
    </div>
  </Layout>
)
```

The other 2 pages for Marin and Sam can now be made with a similar structure. Maybe you are even thinking about another component for the Bio Card!

## Build

With your new Gatsby application complete, it's time to integrate it into your existing HTML website. If you were to build the Gatsby application into static files now and upload them in the place of the existing HTML files, the paths of the links made by Gatsby would not be correct. There are a couple of Gatsby configuration options to fix this.

### Path Prefix

> The `pathPrefix` option in `/gatsby-config.js` tells Gatsby the path at which the build output will be served from. Links to other pages and to the generated JavaScript files that provide [progressive enhancements](/docs/glossary/#progressive-enhancement) for the user will be prefixed with the `pathPrefix` value. This is the config addition for the example Gatsby application, served at the `/who` path:

Reviewer 2: What do you mean by "generated JavaScript files" here? Isn't that covered by pages? "other pages" could also refer to the pages that haven't been touched by Gatsby, and if they were prefixed with `/who`, might not be correct.

```suggestion
The `pathPrefix` option in `/gatsby-config.js` tells Gatsby the path at which the build output will be served from. For example, links to other pages within your Gatsby site will be prefixed with the `pathPrefix` value. Here is some config code to tell Gatsby it should only handle and care about the `/who` path for now:
```

Author: I was trying to acknowledge the many new JavaScript files that we are asking them to host. My thinking is that the audience here will be used to just having a collection of known HTML and CSS files to host. Thinking now that it should be addressed directly with a section if at all though. If assetPrefix is not part of the example I think it's less important to address.

Will pick this up in the discussion below where you brought up assetPrefix.

```js:title=/gatsby-config.js
module.exports = {
  plugins: [`gatsby-plugin-react-helmet`],
  pathPrefix: `/who`, // highlight-line
}
```

### Asset Prefix

> The `assetPrefix` option tells Gatsby about an external location, the asset domain, where its compiled JavaScript files will be served from. In addition to the `pathPrefix`, the link paths for Gatsby's generated JavaScript files will be prefixed with the `assetPrefix` value if provided. In the example website there are three assets already hosted at an asset domain; the style sheets `style.css` and `normalize.css`, and the image `person.png`. If Gatsby's generated JavaScript assets are also to be served from the asset domain, the `assetPrefix` must be set with its address:

Reviewer 2: This seems like overkill to me -- I think it makes more sense to introduce global CSS so people can edit it directly, and we can skip the complexity for asset prefixes (unless it's needed for something specific related to porting within a project, like a pathPrefix is).

```js:title=/gatsby-config.js
module.exports = {
  plugins: [`gatsby-plugin-react-helmet`],
  pathPrefix: `/who`,
  {/* highlight-start */}
  assetPrefix:
    "https://gatsby-html-partial-assets.s3.eu-west-2.amazonaws.com/gatsby",
  {/* highlight-end */}
}
```

### Build step

> You now have a site that mirrors the existing HTML site section. Stop the development server if it's still running, it's time to run the build! 🎉

Reviewer 2: ```suggestion
You now have a site that mirrors the existing HTML site section. Stop the development server if it's still running; it's time to run the production build! 🎉
```

```shell
gatsby build --prefix-paths
```

> **Note**: The `--prefix-paths` option _must_ be used for path and asset prefixes to be applied

Once a build is complete, the compiled set of files can be found in `/public`. It's all in there and ready to replace the existing files! In the case of the example site, the folder contents are deployed directly to the `/who` path of the website domain in place of the existing HTML files. If `assetPrefix` is also being used, the contents are also uploaded to a new `/gatsby/who` folder on the asset domain.

### Integrated site file structure

> Here is the structure of the HTML & non JavaScript asset files after the built, Gatsby `/who` section is added to website-domain and the `/gatsby/who/` folder is added at the asset domain:

Reviewer 3: ```suggestion
Here is the structure of the HTML & non JavaScript asset files after the built, Gatsby `/who` section is added to website domain and the `/gatsby/who/` folder is added at the asset domain:
```

> Here is the structure of the HTML & non JavaScript asset files after the built, Gatsby `/who` section is added to website domain and the `/gatsby/who/` folder is added at the asset domain:

Reviewer 2: ```suggestion
Here is the structure of the HTML & non-JavaScript asset files after the built Gatsby `/who` section is added to website domain and the `/gatsby/who/` folder is added at the asset domain:
```

1. Mirror the website structure within the Gatsby `/src/pages` folder.

2. Adjust the `pathPrefix` Gatsby configuration option to reflect the new root path at which the Gatsby application will be served.

3. Ensure the `to` props on all Gatsby `<Link>` components are correct relative to `/src/pages/` as the application root.

For the example website covered in this guide, here is what each step involves for also migrating the `services` part of the site:

1. Create a `/src/pages/who` folder, and move into it all the files currently in `/src/pages`. Create a new `/src/pages/services` folder and follow the html migration steps above for the `/services/` HTML files. The already completed Layout component will save a lot of the work!

```
gatsby-site-sections/src/pages

├── who
│   ├── index.js
│   ├── ellla-arborist.js
│   ├── marin-leafer.js
│   └── sam-surgeon.js
└── services
    ├── index.js
    ├── growing.js
    ├── cleaning.js
    └── shrinking.js
```

2. The `pathPrefix` configuration option is no longer needed as the Gatsby application will now be served at the root of the site alongside the remaining HTML files.

> 3. Add `/who` to the start of the `to` attributes in the existing index.js, and do the equivalent for `/services`. In the `Layout.js` component file, change the `<a>` tag linking to the services page to a Gatsby `<Link>` component.

Reviewer 2: A short code example would be helpful here.

After following those steps for the example site it is pretty much entirely Gatsby! Migrating the rest of the HTML files, `/about.html`, `/contact.html`, `/404.html` and `/index.html` to Gatsby pages would enable pre-fetching links between all the pages in the site. The complete Gatsby application is ready to take full advantage of Gatsby and it's community.

## Next steps

> Gatsby can handle assets through direct imports to page and component files, the [asset import documentation](/docs/importing-assets-into-files/) covers imports as well as the less optimized static folder. Once assets are handled through Gatsby, plugins can be used to optimize their processing and delivery.

Reviewer 2: ```suggestion
Gatsby can handle assets through direct imports to page and component files; the [asset import documentation](/docs/importing-assets-into-files/) covers imports as well as the less-optimized `static` folder. Once assets are handled through Gatsby, plugins can be used to optimize their processing and delivery.
```

> The [building with components page](/docs/building-with-components/) has information about why Gatsby uses React component architecture and how it fits into a Gatsby application.

Reviewer 2: ```suggestion
The [building with components doc](/docs/building-with-components/) has information about why Gatsby uses React component architecture and how it fits into a Gatsby application.
```

[Sourcing content and data](/docs/content-and-data/) is a great next step if you are interested in separating your content from your website code.

> Short guides can be found at the [recipes section](/docs/recipes), for adding functionality such as optimizing and querying local images, adding markdown support and integrating various modern CSS tools. The [adding website functionality page](/docs/adding-website-functionality/) has longer guides for larger tasks such as making your site accessible, adding authentication and doing client side data fetching.

Reviewer 2: ```suggestion
Short guides can be found at the [recipes section](/docs/recipes) for adding functionality such as optimizing and querying local images, adding Markdown support and integrating various modern CSS tools. The [adding website functionality page](/docs/adding-website-functionality/) has longer guides for larger tasks such as making your site accessible, adding authentication, and fetching data with client-side JavaScript.
```

{/* highlight-end */}
```

## Porting more parts

To replace multiple parts of a website with a single Gatsby application there are a few additional steps:

1. Mirror the website structure within the Gatsby `/src/pages` folder.

2. Adjust the `pathPrefix` Gatsby configuration option to reflect the new root path at which the Gatsby application will be served.

> 3. Ensure the `to` props on all Gatsby `<Link>` components are correct relative to `/src/pages/` as the application root.

Reviewer 3: general question (maybe adressed also early on this page): here we use `<Link>` for all pages in the old and new website or only for the new website (preloading is compatible with old html pages?)

Reviewer 3: i think you adressed this in the next point 3

Author: The [docs](https://www.gatsbyjs.org/docs/gatsby-link/#replace-a-tags-with-the-link-tag-for-local-links) warn against using `<Link>` for external sites, I took that to mean anything outside the Gatsby app but looking back it may just be external to the website as a whole.

Once `pathPrefix` is not being used in a project, I don't know if there is a reason not to use `<Link>` for non Gatsby, internal pages. That is a possible small clarification to add to the [Link page](https://www.gatsbyjs.org/docs/gatsby-link/#replace-a-tags-with-the-link-tag-for-local-links). Will dig a little to find the answer.

For the example website covered in this guide, here is what each step involves for also migrating the `services` part of the site:

1. Create a `/src/pages/who` folder, and move into it all the files currently in `/src/pages`. Create a new `/src/pages/services` folder and follow the html migration steps above for the `/services/` HTML files. The already completed Layout component will save a lot of the work!

```
gatsby-site-sections/src/pages

├── who
│   ├── index.js
│   ├── ellla-arborist.js
│   ├── marin-leafer.js
│   └── sam-surgeon.js
└── services
    ├── index.js
    ├── growing.js
    ├── cleaning.js
    └── shrinking.js
```

> 2. The `pathPrefix` configuration option is no longer needed as the Gatsby application will now be served at the root of the site alongside the remaining HTML files.

Reviewer 3: but we still need the build with  `--withPaths` for the assets?

Author: Aye, there's a note up with the build step on line 479 that I think does the trick in explaining this:

`> **Note**: The `--prefix-paths` option _must_ be used for path and asset prefixes to be applied`

Thinking they'll find themselves back up there when building again if they hit problems.

Reviewer 2: Here's a question: what does the asset prefix stuff gain? Could this guide be simplified and include global CSS instead?

Author: Looking again at the [asset prefix docs](https://www.gatsbyjs.org/docs/asset-prefix/) page as well, I think it'll follow on from this guide really cleanly for the reader. I'll take a pass through to take out asset-prefix from the example, host the whole gatsby build output at the /who path and move the css, png etc to assets folder on the website-domain.

Author: What do you mean by global CSS? Directly importing a CSS file in the Layout.js file? I'd lean more toward keeping a basic link tag so the method stays consistent with the remaining HTML pages, so

`<link
      rel="stylesheet"
      type="text/css"
      href="https://gatsby-html-partial-assets.s3.eu-west-2.amazonaws.com/style.css"
    />`

becomes

`<link rel="stylesheet" type="text/css" href="/assets/style.css"/>`
